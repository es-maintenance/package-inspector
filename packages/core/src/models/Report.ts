import fs from 'fs-extra';
import { IPackage } from './Package';

export type LatestPackages = Record<string, string>;

export interface DependenciesMap {
  [dependencyName: string]: IPackage;
}

export interface SuggestionAction {
  message: string;
  targetPackageId: string;
}

export interface Suggestion {
  id: string;
  pluginTarget: string;
  name: string;
  message: string;
  actions: SuggestionAction[];
}

export interface PluginInfo {
  [pluginKey: string]: any;
}

/**
 * @example
 ```
{
    "latestPackages": { "package-name": "version", "package-2": "version" }, // type { [key: string]: semver }
    "root": {
        // type Package
        "name": "docusaurus-plugin-search-local",
        "version": "0.5.0",
        "metadata": {
            "size": {
                // type SizeInfo
                "files": 1230,
                "physical": 386391
            },
            "pathsOnDisk": [
                // type string[]
                "/Users/lemiller/projects/docusaurus-plugin-search-local"
            ]
        },
        "funding": "N/A",
        "homepage": "N/A",
        "dependencies": ["leet-core@1.33.7"] // type alias dependencyKeyArray = string[]
    },
    "dependencies": {
        // type { [key: string]: Package }
        "@algolia/autocomplete-core@0.5.0": {
            // type Package
            "funding": "N/A",
            "homepage": "N/A",
            "name": "@algolia/autocomplete-core",
            "version": "0.5.0",
            "type": "prod",
            "metadata": {
                "size": {
                    // type SizeInfo
                    "files": 1230,
                    "physical": 386391
                },
                "pathsOnDisk": [
                    // type string[]
                    "/Users/lemiller/projects/docusaurus-plugin-search-local/node_modules/@algolia/autocomplete-core"
                ]
            },
            "dependencies": ["bar-baz@1.5.2", "foo-bar@1.3.7"] // type string[]
        }
    },
    "suggestions": [{
        "id": "packagesWithPinnedVersions",
        "name": "Packages with pinned dependencies",
        "message": "There are currently 38 packages with pinned versions which will never collapse those dependencies causing an additional 5.5 MiB",
        "actions": [{
            "message": "\"terser\" (parcel#@parcel/config-default#@parcel/optimizer-htmlnano#htmlnano#terser) has a pinned version for source-map-support@~0.5.20 that will never collapse.",
            "targetPackageId": "package@version" // type PackageKey
        }]
    }],
    "pluginInfo": {
        "plugin-name@version": {} // type PluginInfo that comes from the plugin
    }
}
 ```
 */

export interface SerializedReport {
  latestPackages: LatestPackages;
}

export function serializeReport(jsonReport: any): SerializedReport {
  const errors: string[] = [];

  const serializedReport: SerializedReport = {
    latestPackages: {},
  };

  if (jsonReport.hasOwnProperty('latestPackages')) {
    if (typeof jsonReport.latestPackages !== 'object') {
      errors.push(
        `"jsonReport.latestPackages" should be of type "object" but found "${typeof jsonReport.latestPackages}"`
      );
    } else {
      for (const [key, value] of Object.entries(jsonReport.latestPackages)) {
        if (typeof value !== 'string') {
          errors.push(
            `"jsonReport.latestPackages['${key}']" should be of type "string" but found "${typeof value}"`
          );
        } else {
          serializedReport.latestPackages[key] = value;
        }
      }
    }
  }

  if (errors.length) {
    throw new Error(
      `Error(s) occurred while processing the provided report:
      ${errors.join('\n')}`
    );
  }

  return serializedReport;
}

export class Report {
  root: IPackage;

  latestPackages: LatestPackages = {};
  dependencies: DependenciesMap = {};
  suggestions: Suggestion[] = [];
  pluginInfo: PluginInfo = {};

  get id(): string {
    return '';
  }

  constructor() {
    this.root = {
      name: '',
      version: '',
      dependencies: [],
    };
  }

  loadFromFile(jsonFilePath: string) {
    const object = fs.readJSONSync(jsonFilePath);

    const serializedReport = serializeReport(object);

    this.root = object.root;
    this.dependencies = object.dependencies;
    this.suggestions = object.suggestions;
    this.pluginInfo = object.pluginInfo;
    this.latestPackages = serializedReport.latestPackages;
  }
}
