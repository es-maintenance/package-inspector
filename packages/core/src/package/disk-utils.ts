import debug from 'debug';
import fs from 'fs';
import path from 'path';

const logger = debug('pi-core:disk-utils');

export function getAllFiles(
  dirPath: string,
  exclude?: RegExp,
  arrayOfFiles?: string[]
) {
  const files = fs.readdirSync(dirPath);

  const _arrayOfFiles = arrayOfFiles || [];

  files.forEach(function (file: string) {
    const fullPath = path.join(dirPath, file);

    try {
      if (
        fs.statSync(fullPath).isDirectory() &&
        (!exclude || (exclude && !exclude.test(fullPath)))
      ) {
        try {
          _arrayOfFiles.push(...getAllFiles(fullPath, exclude));
        } catch (ex: any) {
          logger(ex.message);
        }
      } else {
        if (!exclude || (exclude && !exclude.test(fullPath))) {
          _arrayOfFiles.push(fullPath);
        } else {
          logger('excluding', fullPath, ' for ', dirPath, exclude);
        }
      }
    } catch (ex: any) {
      logger(ex.message);
    }
  });

  return _arrayOfFiles;
}

export function getDirectorySize({
  directory,
  exclude,
}: {
  directory: string;
  exclude?: RegExp;
}): { files: number; physical: number } {
  const arrayOfFiles = getAllFiles(directory, exclude);

  const size = arrayOfFiles
    .map((filePath) => fs.statSync(filePath).size)
    .reduce((a, b) => a + b, 0);

  return {
    files: arrayOfFiles.length,
    physical: size,
  };
}

export function stripPathOnDisk(nodePath: string, workingPath: string): string {
  if (process.env.NODE_ENV === 'test') {
    // macOS uses a sym-link for `/tmp -> /private/tmp` dir and `nodePath` will start `/private`
    const slugToStrip = nodePath.startsWith('/private')
      ? `/private${workingPath}`
      : workingPath;

    return nodePath.replace(slugToStrip, '');
  }

  return nodePath;
}
