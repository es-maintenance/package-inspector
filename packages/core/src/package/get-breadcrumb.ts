import { IArboristNode } from '../types';

export function getBreadcrumb(node: IArboristNode): string {
  const bread: string[] = [];

  function walk(node: IArboristNode): string[] {
    if (bread.includes(node.name)) {
      return bread;
    }

    if (node.edgesIn) {
      const [edge] = [...node.edgesIn.values()];

      if (edge && edge.from) {
        bread.push(edge.name);

        return walk(edge.from);
      } else {
        // we have gotten to the root project, don't push the root project name
        return bread;
      }
    } else {
      return bread;
    }
  }

  return walk(node).reverse().join('#');
}
