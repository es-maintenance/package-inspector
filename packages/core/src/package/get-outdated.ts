import { exec } from 'child_process';
import { promisify } from 'util';

const execPromise = promisify(exec);

export interface OutdatedData {
  [key: string]: {
    latest: string;
  };
}

export async function getOutdated(dir: string): Promise<OutdatedData> {
  let packageData: OutdatedData = {};

  try {
    await execPromise('npm outdated --json', {
      cwd: dir,
    });
  } catch (ex: any) {
    packageData = JSON.parse(ex.stdout.toString('utf8')) as OutdatedData;
  }

  return packageData;
}
