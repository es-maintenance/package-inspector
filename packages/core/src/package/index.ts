export { getAllFiles, getDirectorySize } from './disk-utils';
export { getBreadcrumb } from './get-breadcrumb';
export { getLatestPackages } from './get-latest-packages';
export { humanFileSize } from './human-file-size';
