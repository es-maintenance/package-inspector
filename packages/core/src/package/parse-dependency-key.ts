// @babel/core@0.5.0
// core-js@0.5.0
export function parseDependencyKey(key: string): {
  name: string;
  version?: string;
} {
  const parts = key.split('@');

  // We found an `@` symbol at the beginning of the key, so we have a scoped package name without a version.
  if (parts.length === 2 && parts[0] === '') {
    return {
      name: key,
    };
  }

  // We hit a case like `@babel/core@0.5.0`
  if (parts.length === 3) {
    return {
      name: `${parts[0]}@${parts[1]}`,
      version: parts[2],
    };
  }

  return {
    name: `${parts[0]}`,
    version: parts[1],
  };
}
