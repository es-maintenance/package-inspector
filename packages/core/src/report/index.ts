export { generateReport, type Report } from './generate-report';
