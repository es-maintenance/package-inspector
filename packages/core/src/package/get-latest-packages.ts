import debug from 'debug';
import { copyFile, writeFile } from 'fs/promises';
import { IDependencyMap } from 'package-json-type';
import path from 'path';
import tmp from 'tmp';

import { IArboristNode } from '../types';
import { getOutdated } from './get-outdated';

let _CACHEDLATESTPACKAGES: IDependencyMap = {};

export async function getLatestPackages(
  arboristValues: IArboristNode[]
): Promise<IDependencyMap> {
  // only do this once
  if (Object.keys(_CACHEDLATESTPACKAGES).length) {
    return _CACHEDLATESTPACKAGES;
  }
  // TODO: this should be cached for faster build times and not having to make large requests to NPM
  const fakePackageJson: { dependencies: IDependencyMap } = {
    dependencies: {},
  };

  const dependencyKeys = arboristValues
    .filter((node) => {
      // ignore linked packages and packages that have realpaths that are on disk (which means they are linked and potentially don't exist in the registry)
      return (
        (!node.isLink || !node.isWorkspace) &&
        node.realpath.includes('node_modules')
      );
    })
    .map((node) => node.name);

  dependencyKeys.forEach((key) => {
    if (key.includes('fastlane')) return;
    fakePackageJson.dependencies[key] = '*';
  });

  const tmpobj = tmp.dirSync({ unsafeCleanup: true });
  debug('Dir: ' + tmpobj.name);

  await writeFile(
    path.resolve(tmpobj.name, 'package.json'),
    JSON.stringify(fakePackageJson)
  );

  const latestPackages: IDependencyMap = {};
  try {
    // need to keep the registry context for the fake package.json so things get resolved correctly when checking outdated
    await copyFile(
      path.resolve(process.cwd(), '.npmrc'),
      path.resolve(tmpobj.name, '.npmrc')
    );
  } catch (ex) {
    debug('Could not copy .npmrc, using the default npm registry.');
  }

  const outdatedPackages = await getOutdated(tmpobj.name);

  for (const packageName in outdatedPackages) {
    latestPackages[packageName] = outdatedPackages[packageName].latest;
  }

  tmpobj.removeCallback();
  _CACHEDLATESTPACKAGES = latestPackages;

  return latestPackages;
}
