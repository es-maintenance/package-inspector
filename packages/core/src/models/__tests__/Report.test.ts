import {
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  SpyInstance,
  test,
  vi,
} from 'vitest';
import fixturify, { type DirJSON } from 'fixturify';
import fse from 'fs-extra';
import tmp from 'tmp';
import path from 'path';
import createDebug from 'debug';

import * as GetOutdatedModule from '../../package/get-outdated';
import { Report, serializeReport } from '../Report';

import NO_NODE_MODULES from '../../__tests__/__fixtures__/no-node-modules.json';
import MIN_NODE_MODULES from '../../__tests__/__fixtures__/min-node-modules.json';
import { generateReport } from '../../report';

describe('Report', () => {
  const logger = createDebug('pi-core:Report:Test');

  let cleanupTmpDir = () => {};
  let getOutdatedSpy: SpyInstance;
  let tmpLocation: string;
  let __tmp_dir__: string;

  beforeAll(() => {
    tmp.setGracefulCleanup();

    __tmp_dir__ = path.join(__dirname, '__tmp__');

    fse.ensureDirSync(__tmp_dir__);

    logger(`Using tmpDir: ${__tmp_dir__}`);
  });

  beforeEach(() => {
    const tmpObject = tmp.dirSync({
      tmpdir: __tmp_dir__,
      unsafeCleanup: true,
    });

    tmpLocation = tmpObject.name;
    cleanupTmpDir = tmpObject.removeCallback;
  });

  afterEach(() => {
    cleanupTmpDir();
  });

  describe('loadFromFile', () => {
    test('it should load a simple json report', async () => {
      // Write a dummy project structure
      fixturify.writeSync(tmpLocation, NO_NODE_MODULES as DirJSON);

      // Do not have any outdated packages
      getOutdatedSpy = vi
        .spyOn(GetOutdatedModule, 'getOutdated')
        .mockImplementation(async () => ({}));

      // Generate a report based off of dummy project structure.
      const generatedReport = await generateReport(tmpLocation);

      // Write the report to disk
      const reportPath = path.join(tmpLocation, 'report.json');
      fse.writeFileSync(reportPath, JSON.stringify(generatedReport));

      // Create a new Report instance.
      const report = new Report();

      // Verify the structure of the initial Report instance.
      expect(report).toMatchSnapshot();

      // Serialize the report from a file.
      report.loadFromFile(reportPath);

      // Verify the populated Report instance
      expect(report).toMatchSnapshot();
    });

    test('it should load a json report with minimal node modules', async () => {
      // Write a dummy project structure
      fixturify.writeSync(tmpLocation, MIN_NODE_MODULES as DirJSON);

      // Do not have any outdated packages
      getOutdatedSpy = vi
        .spyOn(GetOutdatedModule, 'getOutdated')
        .mockImplementation(async () => ({}));

      // Generate a report based off of dummy project structure.
      const generatedReport = await generateReport(tmpLocation);

      // Write the report to disk
      const reportPath = path.join(tmpLocation, 'report.json');
      fse.writeFileSync(reportPath, JSON.stringify(generatedReport));

      // Create a new Report instance.
      const report = new Report();

      // Verify the structure of the initial Report instance.
      expect(report).toMatchSnapshot();

      // Serialize the report from a file.
      report.loadFromFile(reportPath);

      // Verify the populated Report instance
      expect(report).toMatchSnapshot();
    });
  });

  describe('serializeReport', () => {
    test('it should serialize a serialize a simple report', async () => {
      // Write a dummy project structure
      fixturify.writeSync(tmpLocation, NO_NODE_MODULES as DirJSON);

      // Do not have any outdated packages
      getOutdatedSpy = vi
        .spyOn(GetOutdatedModule, 'getOutdated')
        .mockImplementation(async () => ({}));

      // Generate a report based off of dummy project structure.
      const generatedReport = await generateReport(tmpLocation);

      // Write the report to disk
      const reportPath = path.join(tmpLocation, 'report.json');
      fse.writeFileSync(reportPath, JSON.stringify(generatedReport));

      // Read report from disk
      const jsonReport = JSON.parse(fse.readFileSync(reportPath, 'utf-8'));

      // Create a serialized Report;
      const report = serializeReport(jsonReport);

      // Verify the simple serialized report.
      expect(report).toMatchSnapshot();
    });

    test('it should serialize a project with minimal node modules', async () => {
      // Write a dummy project structure
      fixturify.writeSync(tmpLocation, MIN_NODE_MODULES as DirJSON);

      // Do not have any outdated packages
      getOutdatedSpy = vi
        .spyOn(GetOutdatedModule, 'getOutdated')
        .mockImplementation(async () => ({}));

      // Generate a report based off of dummy project structure.
      const generatedReport = await generateReport(tmpLocation);

      // Write the report to disk
      const reportPath = path.join(tmpLocation, 'report.json');
      fse.writeFileSync(reportPath, JSON.stringify(generatedReport));

      // Read report from disk
      const jsonReport = JSON.parse(fse.readFileSync(reportPath, 'utf-8'));

      // Create a serialized Report;
      const report = serializeReport(jsonReport);

      // Verify the minimal serialized report.
      expect(report).toMatchSnapshot();
    });

    test('it should serialize "latestPackages"', async () => {
      // Write a dummy project structure
      fixturify.writeSync(tmpLocation, NO_NODE_MODULES as DirJSON);

      // Do not have any outdated packages
      getOutdatedSpy = vi
        .spyOn(GetOutdatedModule, 'getOutdated')
        .mockImplementation(async () => ({
          foo: {
            latest: '1.33.7',
          },
          bar: {
            latest: '7.33.1',
          },
        }));

      // Generate a report based off of dummy project structure.
      const generatedReport = await generateReport(tmpLocation);

      // Write the report to disk
      const reportPath = path.join(tmpLocation, 'report.json');
      fse.writeFileSync(reportPath, JSON.stringify(generatedReport));

      // Read report from disk
      const jsonReport = JSON.parse(fse.readFileSync(reportPath, 'utf-8'));

      // Create a serialized Report;
      const report = serializeReport(jsonReport);

      // Verify the simple serialized report.
      expect(report).toMatchSnapshot();
    });

    test('it should error out during serialization when incorrect "latestPackages" is found', () => {
      // The base `latestPackages` field is incorrect;
      expect(() =>
        serializeReport({
          latestPackages: '',
        })
      ).toThrowError(
        `"jsonReport.latestPackages" should be of type "object" but found "string"`
      );

      // `latestPackages['foo']` is of type number when it should be a semver string
      expect(() =>
        serializeReport({
          latestPackages: {
            foo: 123,
          },
        })
      ).toThrowError(
        `"jsonReport.latestPackages['foo']" should be of type "string" but found "number"`
      );
    });
  });
});
