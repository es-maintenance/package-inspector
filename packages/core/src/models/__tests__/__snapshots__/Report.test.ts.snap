// Vitest Snapshot v1

exports[`Report > loadFromFile > it should load a json report with minimal node modules 1`] = `
Report {
  "dependencies": {},
  "latestPackages": {},
  "pluginInfo": {},
  "root": {
    "dependencies": [],
    "name": "",
    "version": "",
  },
  "suggestions": [],
}
`;

exports[`Report > loadFromFile > it should load a json report with minimal node modules 2`] = `
Report {
  "dependencies": {
    "left-pad@1.3.0": {
      "dependencies": [],
      "metadata": {
        "pathsOnDisk": [
          "/node_modules/left-pad",
        ],
        "size": {
          "files": 10,
          "physical": 9752,
        },
      },
      "name": "left-pad",
      "type": "prod",
      "version": "1.3.0",
    },
  },
  "latestPackages": {},
  "pluginInfo": {},
  "root": {
    "dependencies": [],
    "metadata": {
      "pathsOnDisk": [
        "",
      ],
    },
    "name": "min-node-modules",
    "version": "1.0.0",
  },
  "suggestions": [
    {
      "actions": [],
      "id": "packagesWithPinnedVersions",
      "message": "There are currently 0 pinned dependencies",
      "name": "Packages with pinned dependencies",
    },
    {
      "actions": [],
      "id": "packagesWithExtraArtifacts",
      "message": "There are currently 0 extra artifacts",
      "name": "Packages with extra artifacts",
    },
    {
      "actions": [],
      "id": "notBeingAbsorbedByTopLevel",
      "message": "There are currently 0 duplicate packages being installed on disk",
      "name": "Dependencies not being absorbed",
    },
    {
      "actions": [],
      "id": "nestedDependencyFreshness",
      "message": "Out of the total 1 sub packages currently installed; 0 major versions out of date (0.00%), 0 minor versions out of date (0.00%), 0 patch versions out of date (0.00%)",
      "name": "Nested Dependency Freshness",
    },
    {
      "actions": [],
      "id": "topLevelDepsFreshness",
      "message": "Out of the total 1 explicit dependencies defined in the package.json; 0 major versions out of date (0.00%), 0 minor versions out of date (0.00%), 0 patch versions out of date (0.00%)",
      "name": "Top Level Dependency Freshness",
    },
  ],
}
`;

exports[`Report > loadFromFile > it should load a simple json report 1`] = `
Report {
  "dependencies": {},
  "latestPackages": {},
  "pluginInfo": {},
  "root": {
    "dependencies": [],
    "name": "",
    "version": "",
  },
  "suggestions": [],
}
`;

exports[`Report > loadFromFile > it should load a simple json report 2`] = `
Report {
  "dependencies": {},
  "latestPackages": {},
  "pluginInfo": {},
  "root": {
    "dependencies": [],
    "metadata": {
      "pathsOnDisk": [
        "",
      ],
    },
    "name": "no-node-modules",
    "version": "1.0.0",
  },
  "suggestions": [
    {
      "actions": [],
      "id": "packagesWithPinnedVersions",
      "message": "There are currently 0 pinned dependencies",
      "name": "Packages with pinned dependencies",
    },
    {
      "actions": [],
      "id": "packagesWithExtraArtifacts",
      "message": "There are currently 0 extra artifacts",
      "name": "Packages with extra artifacts",
    },
    {
      "actions": [],
      "id": "notBeingAbsorbedByTopLevel",
      "message": "There are currently 0 duplicate packages being installed on disk",
      "name": "Dependencies not being absorbed",
    },
    {
      "actions": [],
      "id": "nestedDependencyFreshness",
      "message": "Out of the total 0 sub packages currently installed; 0 major versions out of date (NaN%), 0 minor versions out of date (NaN%), 0 patch versions out of date (NaN%)",
      "name": "Nested Dependency Freshness",
    },
    {
      "actions": [],
      "id": "topLevelDepsFreshness",
      "message": "Out of the total 0 explicit dependencies defined in the package.json; 0 major versions out of date (NaN%), 0 minor versions out of date (NaN%), 0 patch versions out of date (NaN%)",
      "name": "Top Level Dependency Freshness",
    },
  ],
}
`;

exports[`Report > serializeReport > it should serialize "latestPackages" 1`] = `
{
  "latestPackages": {
    "bar": "7.33.1",
    "foo": "1.33.7",
  },
}
`;

exports[`Report > serializeReport > it should serialize a project with minimal node modules 1`] = `
{
  "latestPackages": {},
}
`;

exports[`Report > serializeReport > it should serialize a serialize a simple report 1`] = `
{
  "latestPackages": {},
}
`;
